Updates influx/questdb comp blog (#1829)
questdb/questdb.io: blog/2023-04-04-comparing-questdb-timescaledb-influxdb.mdx

  width={650}
></Banner>

> _Comparison courtesy of regular guest contributor [Yitaek Hwang](https://yitaek.medium.com/)._

Reviewer: Should we get rid of this too, and thus have a cleaner intro that starts with a header, without the prelude about yitaek?

  width={650}
></Banner>

As application use cases have expanded and accelerated, developers have outgrown the capabilities of conventional relational databases. New databases have emerged with solutions to urgent and emerging requirements. Among them, no requirement has grown more than the need to handle explosive volumes of **time-series data**: 

> 

Reviewer: We fail to say what the screenshot is about. It tracks the popularity of time-series databases

While other databases can also handle time-series data to a certain extent,
TSDBs are designed with the above properties to handle data ingestion, storage,
and aggregation over time more efficiently. So as the demand for time-series
data continues to explode on the heels of cloud computing, IoT, financial market
data, and machine learning, how should architects go about choosing a
time-series database?

## InfluxDB

First released in 2013, InfluxDB is the market leader in the TSDB space,
overtaking previous incumbents Graphite and OpenTSDB. As with many OSS database companies, [InfluxDB](https://www.influxdata.com/) is licensed with MIT License for a single node, with paid-plans available for InfluxDB Cloud and InfluxDB enterprise that provide clustering and other production-ready features.

<Screenshot
  alt="A table showing the current popularity of time-series databases as shown by db-engines.com"
  height={591}
  src="/img/blog/2023-04-04/2023-time-series-popularity-log-scale.webp"
  width={770}

>   title="Image Credit: DB-engines"

Reviewer: Could we be more descriptive in the image title, to explain what we are displaying

  your applications. In any case, it is possible to enforce schemas for specific
  buckets if needed.

- **Data storage (TSM & TSI):** InfluxDB stores data points in time-structured merge
  tree (TSM) and time-series index (TSI) files. As a result of the split between their storage model and an additional TSI index, deviations from the preferred schema pattern can lead to degraded performance. 
  
  The primary example involves [high cardinality](/glossary/high-cardinality/) data, where rows within a column contain many unique or varying elements. In InfluxDB, arbitrary tags are tied to a specialized "meta index" (TSI). These tags join with a measurement and field key to create a series, which is then stored in memory to be read. As a result, more tags means more series stored in memory, which leads to higher - and more volatile - memory requirements. If the cardinality is variable, so too is the pattern in which CPU and memory is utilized, making it more difficult to anticipate load.
  

>   While InfluxDB Cloud has implemented an updated columnar model to remove the additional TSI index, it remains in the open-source version which still struggles with higher cardinality data. As evidence, in InfluxDB OSS 2.6, the [schema design documentation](https://docs.influxdata.com/influxdb/v2.7/write-data/best-practices/schema-design/) still offers guidance to _prevent_ high series cardinality.

Reviewer: this section now looks good. Could we get confirmation from Andrei about this specific bit?

 **Data storage (TSM & TSI):** InfluxDB stores data points in time-structured merge
  tree (TSM) and time-series index (TSI) files. As a result of the split between their storage model and an additional TSI index, deviations from the preferred schema pattern can lead to degraded performance

Reviewer 2: I would blame TSI index on its own. After all, storing even a few million records in the index isn't a big deal. To me, the biggest contribution to performance degradation with the growth of the number of time series should be disk related. If you store individual time series separately (in individual files or blocks within the same file, doesn't matter much), your disk operations pattern change with the number of time series: when there are only a few of them, the pattern is almost sequential while when there are millions of them it becomes random from the disk perspective. Modern SSD still favor sequential operations for a number of reasons.

Of course, I may be totally wrong here and I didn't profile InfluxDB itself to determine the exact bottlenecks. Nevertheless, I'm fine with the current wording as it's general enough.

Author: Reworking to trim the Influx section down overall, and to be even more general. The point re: cardinality is the key, and we can make it the strongest by simply showing their schema preferences of reduced cardinality.

  buckets if needed.

- **Data storage (TSM & TSI):** data points are stored in time-structured merge
  tree (TSM), and time-series index (TSI) files. [Storage internals](https://docs.influxdata.com/influxdb/v2.6/reference/internals/storage-engine/) are complicated. In short, tuning `component_x` in your primary storage engine might be better for 90% of cases, but create challenges in the remaining 10%. 
  

>   An example is [high cardinality](/glossary/high-cardinality/) data, where rows within a column contain many unique or varying elements. In such a case, overall performance can decrease as optimizations wrought from consistency within column rows can no longer be leveraged. Within InfluxDB, TSI supplements TSM to help manage high cardinality data, which raises the cumulative under-the-hood complexity to offset the challenge.

Reviewer: I would like @javier and @puzpuzpuz to have a look at this paragraph rewrite

Reviewer 2: The change looks fine to me. The previous text was indeed a bit confusing.

> Within InfluxDB, TSI supplements TSM to help manage high cardinality data, which raises the cumulative under-the-hood complexity to offset the challenge.

I don't think under-the-hood complexity is relevant for the end user: as long as the ingestion and query performance is good enough, users are happy. We should stress that high cardinality is problematic for Influx v1 & v2 since they store each time series separately, hence (almost) random disk read and write pattern appears.

Author: Reworked the section after more research and considering Andrei's feedback.

It lead to an additional con, too, which is that the Cloud offering seems to have taken a large step here that isn't present in the OSS version.

Reviewer 2: The updated version looks good.

<Screenshot
  alt="A diagram showing TimescaleDB and InfluxDB on two ends of the SQL vs NoSQL spectrum."
  height={591}
  src="/img/blog/2023-04-04/sql-v-nosql.jpeg"
  width={770}
  title="TimescaleDB is based on PostgreSQL, InfluxDB its own paradigm - they are at opposite ends of the SQL spectrum. Image credit: TimescaleDB."
/>

TimescaleDB is built upon the traditional row-based architecture of PostgreSQL, as opposed to the modernized column-based architecture seen in DuckDB, InfluxDB Cloud and QuestDB. Row-based architecture performs well within Online Transaction Processing (OLTP) where use cases see a large number of simple inserts, updates and deletes by many users. By comparison, columnar models excel within Online Analytical Processing (OLAP) cases, which require much higher ingestion speeds and more complex queries, exemplified by the analysis of market or systems data. 

While performance lags behind competitors, TimescaleDB has improved upon the OLTP foundation of PostreSQL with the following features:

> - **Hypertables:** Hypertables sit alongside standard PostgreSQL tables to better handle time-series data. They require that data [partition into configurable chunks](https://docs.timescale.com/timescaledb/latest/overview/core-concepts/hypertables-and-chunks/#partitioning-in-hypertables-with-chunks) based on a time column. As a result, users must consider chunk memory handling, compression, data re-ordering, and the replication of transactions across nodes, to achieve optimal performance.

Author: ```suggestion
- **Hypertables:** Hypertables sit alongside standard PostgreSQL tables to better handle time-series data. They require data to be [partitioned into configurable chunks](https://docs.timescale.com/timescaledb/latest/overview/core-concepts/hypertables-and-chunks/#partitioning-in-hypertables-with-chunks) based on a time column. As a result, users must consider chunk memory handling, compression, data re-ordering, and the replication of transactions across nodes, to achieve optimal performance.
```

> - **Hypertables:** Hypertables sit alongside standard PostgreSQL tables to better handle time-series data. They require data to be [partitioned into configurable chunks](https://docs.timescale.com/timescaledb/latest/overview/core-concepts/hypertables-and-chunks/#partitioning-in-hypertables-with-chunks) based on a time column. As a result, users must consider chunk memory handling, compression, data re-ordering, and the replication of transactions across nodes, to achieve optimal performance.

Author: ```suggestion
- **Hypertables:** Hypertables sit alongside standard PostgreSQL tables to better handle time-series data. They require data to be [partitioned into configurable chunks](https://docs.timescale.com/timescaledb/latest/overview/core-concepts/hypertables-and-chunks/#partitioning-in-hypertables-with-chunks) based on a time column. Chunks need to be tuned for optimal performance.
```

>   In the latest (non open source) InfluxDB Cloud IOx,

Reviewer: influx section is quite long.

Shall we remove " Still considering the large community and
  [integrations](https://www.influxdata.com/products/integrations/) that
  InfluxDB has assembled, some of the advantages of Flux start to materialize,
  especially when combined with built-in dashboards." ?

## QuestDB

> QuestDB is an open-source TSDB written in low-latency Java, C++, and Rust. It ranks in \#10 on [DB-Engines Time Series DBMS](https://db-engines.com/en/ranking/time+series+dbms),

Reviewer: Replace the image with the newer version that shows a time series for timescale,influx and questdb over time

Author: Replaced DB-engines image under influx section 👍
